### ISSUE TITLE
Missing documentation about how URL/queries are quoted/unquoted

### ISSUE THREAD

**Issue by cateee** [pallets/flask#2183]
The help about `flask.request` http://flask.pocoo.org/docs/0.12/api/#incoming-request-data just includes a simple: _"args: A MultiDict with the parsed contents of the query string. (The part in the URL after the question mark)."_, Thus nothing about how args are generated.

I need to pass a complex string (with spaces and other url characters).

I found from some experiments, ithat I can generate on client side the query string with `urllib.parse.quote(query1, safe='')`, and this query parameter will be unquoted correctly by flask (request.args).

Because there are many quoting convention (e.g. space replaced with '+', it would be nice to have some more info on the documentation about quoting method, and possibly to have the quote and unquote also directly as flask utilities (like the json support), ready to be used client side.

Possibly also the path decoding could have similar problem (in that case with the default 'safe' argument).

**Comment by untitaker** [comment]
>Thus nothing about how args are generated.

Query parameters are defined by the HTTP RFCs. What else is there to document?

**Comment by cateee** [comment]
I searched a lot in the RFCs, and I had no luck. The encoding must be coordinated by client and server.
As far I seen it, they explains the valid transmitted "octets", but nothing about the encoding. This behaviour is frequent on RFC: full character support needs encoding, and it should be specified by an higher protocol.

Note that python specifies two classical methods: https://docs.python.org/3.6/library/urllib.parse.html#url-quoting (one with replaces spaces with + (this one it is used by google in searches) and a second one which also spaces are encoded (it seems your method).

Additional note: I assume that codes in %xx are UTF-8 in your decoding, and HTTP predates UTF-8.

**Comment by untitaker** [comment]
It does seem that the RFCs define very little about this, but https://url.spec.whatwg.org/#application/x-www-form-urlencoded still describes the defaults that Werkzeug mostly follows (though I think Werkzeug partially predates WHATWG). It is probably sufficient to reference just that.

**Comment by untitaker** [comment]
Ah in fact https://www.w3.org/TR/html401/interact/forms.html#h-17.13.4.1 does specify the plus-encoding part.

**Comment by cateee** [comment]
You are right. I went also on that link before opening the issues, but I dismissed it as <form> and POST, forgetting the FORM/GET, so a standardized GET.

So I close the issue (and this bug report could help other clueless people like me, to find the answer).

The answer: to encode urls: uses [`urllib.parse.urlencode`](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlencode), with default `quote` (`urlib.parse.quoteplus`) and default `safe`.
